// @celo/contractkit/lib/wrappers/FeeCurrencyWhitelistWrapper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeeCurrencyWhitelistWrapper = void 0;
require("bignumber.js");
const BaseWrapper_1 = require("./BaseWrapper");
const MINIMAL_TOKEN_INFO_ABI = [
    {
        type: 'function',
        stateMutability: 'view',
        outputs: [{ type: 'string', name: '', internalType: 'string' }],
        name: 'symbol',
        inputs: [],
    },
    {
        type: 'function',
        stateMutability: 'view',
        outputs: [{ type: 'string', name: '', internalType: 'string' }],
        name: 'name',
        inputs: [],
    },
    {
        type: 'function',
        stateMutability: 'view',
        inputs: [],
        outputs: [{ type: 'address', name: '', internalType: 'address' }],
        name: 'adaptedToken',
    },
];
/**
 * FeeCurrencyWhitelist contract listing available currencies usable to pay fees
 */
class FeeCurrencyWhitelistWrapper extends BaseWrapper_1.BaseWrapper {
    constructor() {
        super(...arguments);
        this.getWhitelist = (0, BaseWrapper_1.proxyCall)(this.contract.methods.getWhitelist, undefined, (addresses) => [...new Set(addresses)].sort());
        this.removeToken = (0, BaseWrapper_1.proxyCall)(this.contract.methods.removeToken);
        this.addToken = (0, BaseWrapper_1.proxyCall)(this.contract.methods.addToken);
    }
    getFeeCurrencyInformation(whitelist) {
        return __awaiter(this, void 0, void 0, function* () {
            const feeCurrencies = whitelist !== null && whitelist !== void 0 ? whitelist : (yield this.getWhitelist());
            return Promise.all(feeCurrencies.map((address) => __awaiter(this, void 0, void 0, function* () {
                // @ts-expect-error abi typing is not 100% correct but works
                let contract = new this.connection.web3.eth.Contract(MINIMAL_TOKEN_INFO_ABI, address);
                const adaptedToken = (yield contract.methods
                    .adaptedToken()
                    .call()
                    .catch(() => undefined));
                if (adaptedToken) {
                    // @ts-expect-error abi typing is not 100% correct but works
                    contract = new this.connection.web3.eth.Contract(MINIMAL_TOKEN_INFO_ABI, adaptedToken);
                }
                return Promise.all([
                    contract.methods
                        .name()
                        .call()
                        .catch(() => undefined),
                    contract.methods
                        .symbol()
                        .call()
                        .catch(() => undefined),
                ]).then(([name, symbol]) => ({
                    name,
                    symbol,
                    address,
                    adaptedToken,
                }));
            })));
        });
    }
}
exports.FeeCurrencyWhitelistWrapper = FeeCurrencyWhitelistWrapper;
